### PR TITLE
Fix release resources race that leads to extra worker launches

### LINKED ISSUE #16025: [Streaming shuffle] Workers are started unlimitedly if each ray object is 100KB.
### What is the problem?

It seems like in the reduce phase, workers are started unlimitedly, if each of ray object is 100KB.

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

```
python -m ray.experimental.shuffle --ray-address="localhost:6379" --num-partitions=1000 --partition-size=100e6
```

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by ericl**
Fixes https://github.com/ray-project/ray/issues/16025

This turns out to be a race condition in the initial implementation of avoiding releasing resources on ray.get(). The problem is that when an ObjectRef is deserialized, we are asynchronously resolving its status. This takes an RPC round trip to the owner (0-1ms). If we try to ray.get() an object during this period, then it can release resources even if the object is already created.

This PR fixes the issue by inlining the object status into the reference. This is the same status that would be retried asynchronously later. This ensures:
1. If the status is already available, we don't release resources trying to get the object.
2. We save an extra get status RPC.

### CODE REVIEW

**Review by rkooo567 — APPROVED** [review]
LGTM. Let me organize if I understand things correctly.

Overhead only happens upon ray.get on ref from non-owner.
1. normal case -> 0~1ms until the status is resolved.
2. worst case -> 10ms ray.get block + unnecessary resource release when the object status resolution is slow (extremely uncommon)

Is this correct?

Also, as an alternative solution, we can probably wait indefinitely until the status is resolved (without the grace period)? Isn't the object status supposed be resolved quickly from the owner? (it is not long polling right?)

**Comment by ericl** [comment]
> Also, as an alternative solution, we can probably wait indefinitely until the status is resolved (without the grace period)? Isn't the object status supposed be resolved quickly from the owner? (it is not long polling right?)

The current resolve call blocks until the object is available. We could add a "resolve immediately" RPC which could return if not currently available, and that could be faster than waiting 10ms. However, it would add extra overheads sending the extra RPCs.

**Comment by rkooo567** [comment]
I guess then the 10ms scenario is not "uncommon" in a large DAG? (since it waits until the object is available). But the overhead doesn't seem to be a big deal right now as you mentioned!

**Comment by ericl** [comment]
The 10ms scenario would only occur if we actually ray.get() in the middle
of a DAG. I don't think this scenario is significantly impacted anyways
since releasing resources leads to at minimum a couple hundred milliseconds
to start a new worker. An extra small delay will be lost in that noise.

**Comment by rkooo567** [comment]
Oh yeah. I meant ray.get in the middle of the large dag (sorry for unclear
explanation). But the reasoning makes a lot of sense!

**Comment by ericl** [comment]
Hmm, a significant number of tests are failing on OSX, likely due to slowdowns.

An alternative that might be less overhead:
- When serializing and promoting and object ref to plasma, check if it's created. If so, also store the location data in the serialized ref.
- Upon deserializing such an object ref, we can immediately create the placeholder object without needing an extra resolve future RPC.
This should reduce the GetObjectStatus traffic significantly in many cases, and also this race condition.

**Comment by ericl** [comment]
Changed to a new inlining status approach, which seems to also fix the issue and should perform better. cc @clarkzinzow

File: src/ray/core_worker/future_resolver.h [hunk 1]
```diff
@@ -51,6 +51,15 @@ class FutureResolver {
   /// future.
   void ResolveFutureAsync(const ObjectID &object_id, const rpc::Address &owner_address);
 
+  /// Process a resolved future. This can be used if we already have the objec
```
Comment (clarkzinzow) [inline src/ray/core_worker/future_resolver.h:54]:
```suggestion
  /// Process a resolved future. This can be used if we already have the object
```

**Review by clarkzinzow — APPROVED** [review]
LGTM, this would be an awesome optimization even if it didn't fix the race condition!

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc [hunk 2]
```diff
@@ -154,6 +154,22 @@ void CoreWorkerMemoryStore::GetAsync(
...
+std::shared_ptr<RayObject> CoreWorkerMemoryStore::GetIfExists(const ObjectID &object_id) {
+  std::shared_ptr<RayObject> ptr;
+  {
+    absl::MutexLock lock(&mu_);
+    auto iter = objects_.find(object_id);
+    if (iter != objects_.end()) {
+      ptr = iter->second;
+    }
+  }
+  // It's important for performance to run the callback outside the lock.
```
Comment (rkooo567) [inline src/ray/core_worker/store_provider/memory_store/memory_store.cc:166]: Hmm isn't this still prone to memory corruption? Why don't we make the bool for this function atomic?

**Review by rkooo567 — APPROVED** [review]
LGTM!

**Comment by ericl** [comment]
Hmm some odd test failures.

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc (hunk 2 above)
Comment (ericl) [reply src/ray/core_worker/store_provider/memory_store/memory_store.cc:166]: Yeah good point, let me make another PR to fix this existing issue.

**Comment by ericl** [comment]
Same mac build timeout as others...

**Pull request merged**